src/main/java/com/amazon/opendistroforelasticsearch/sql/rewriter/nestedfield/NestedFieldProjection.java

```diff
@@ -123,23 +217,59 @@ private void buildInnerHit(List<String> fieldNames, NestedQueryBuilder query) {
      * Why linear search? Because NestedQueryBuilder hides "path" field from any access.
      * Assumption: collected NestedQueryBuilder list should be very small or mostly only one.
      */
-    private NestedQueryBuilder findNestedQueryWithSamePath(List<NestedQueryBuilder> nestedQueries, String path) {
+    private NestedQueryBuilder findNestedQueryWithSamePath(
+            List<NestedQueryBuilder> nestedQueries, String path) {
+
         return nestedQueries.stream().
-                             filter(query -> isSamePath(path, query)).
-                             findAny().
-                             orElseGet(createEmptyNestedQuery(path));
+                filter(query -> isSamePath(path, query)).
+                peek(x -> setLastPath(path)).
+                findAny().
+                orElseGet(createNestedQuery(path));
     }
 
     private boolean isSamePath(String path, NestedQueryBuilder query) {
         return nestedQuery(path, query.query(), query.scoreMode()).equals(query);
     }
 
+    /**
+     * groupFieldNamesByPath() uses TreeMap to maintain reverse alphabetical order. Therefore we use previous
+     * path to compare with current path, to help decide if we are done with keeping wrapping nested query
+     * to current nested query.
+     * */
+    private boolean isParentPath(String path, QueryBuilder query) {
+        // query could also be termQueryBuilder, where condition will generate termQueryBuilder
+        if (lastPath.isEmpty() || !(query instanceof NestedQueryBuilder)) {
+            return false;
+        } else {
+            int lastSeparatorIndex = lastPath.lastIndexOf(SEPARATOR);
```

Comment:
np: seems like these two lines are duplicate from the code in Identifier.java. Maybe move these utility method to a separate class and use it from both places?